prov/sockets/src/sock_progress.c

```diff
@@ -2333,29 +2332,27 @@ void sock_pe_remove_rx_ctx(struct sock_rx_ctx *rx_ctx)
 static int sock_pe_progress_rx_ep(struct sock_pe *pe, struct sock_ep_attr *ep_attr,
 					struct sock_rx_ctx *rx_ctx)
 {
-	int ret = 0, i, fd, num_fds;
+	int ret = 0, i, num_fds;
 	struct sock_conn *conn;
 	struct sock_conn_map *map;
+	void *ep_contexts[SOCK_EPOLL_WAIT_EVENTS];
 
 	map = &ep_attr->cmap;
 
-        if (!map->used)
-                return 0;
+	if (!map->used)
+		return 0;
 
-        num_fds = sock_epoll_wait(&map->epoll_set, 0);
-        if (num_fds < 0 || num_fds == 0) {
-                if (num_fds < 0)
-                        SOCK_LOG_ERROR("poll failed: %s\n", strerror(errno));
-                return num_fds;
-        }
+	num_fds = fi_epoll_wait(map->epoll_set, ep_contexts,
+			SOCK_EPOLL_WAIT_EVENTS, 0);
+	if (num_fds < 0 || num_fds == 0) {
+		if (num_fds < 0)
+			SOCK_LOG_ERROR("poll failed: %s\n", strerror(errno));
+		return num_fds;
+	}
 
 	fastlock_acquire(&map->lock);
 	for (i = 0; i < num_fds; i++) {
-		fd = sock_epoll_get_fd_at_index(&map->epoll_set, i);
-		if (fd == -1) /* failed to lookup fd due to connection failures */
-			continue;
-
-		conn = ofi_idm_lookup(&ep_attr->conn_idm, fd);
+		conn = ep_contexts[i];
```

Comment:
I'm referring to an issue in the socket provider code.  Here's the updated code (abbreviated):
```
    num_fds = fi_epoll_wait(map->epoll_set, ep_contexts,
         SOCK_EPOLL_WAIT_EVENTS, 0);

    fastlock_acquire(&map->lock);
    for (i = 0; i < num_fds; i++) 
        ...
```
num_fds will be capped at SOCK_EPOLL_WAIT_EVENTS, which is set to 32.  If the actual number of fd's associated with the epoll set is > 32, we can lose progressing those fd's.  E.g. say there are 40 fd's and all have events.  fi_epoll_wait can only return events for 32 of those.  The other 8 fd's will not be processed as part of the for loop that follows fi_epoll_wait.

We effectively need 2 loops.  The inner loop as given, and the outer loop to retest whenever the number of events returned is equal to the maximum that we request.